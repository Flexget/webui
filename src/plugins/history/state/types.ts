import { Direction } from 'utils/query';

export interface GetHistoryOptions {
  page?: number;
  sortBy?: string;
  order?: Direction.Desc;
}

export interface History {
  task: string;
  title: string;
  url: string;
  filename: string;
  details: string;
  time: string;
  id: number;
}
