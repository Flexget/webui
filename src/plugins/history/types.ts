import { Direction } from 'utils/query';

export const enum GroupByFields {
  Task = 'task',
  Time = 'time',
}

export const enum SortByFields {
  Task = 'task',
  Time = 'time',
  Url = 'url',
  Details = 'details',
  Id = 'id',
  Filename = 'filename',
  Title = 'title',
}

export interface GetHistoryOptions {
  page?: number;
  sortBy?: SortByFields;
  order?: Direction;
  task?: string;
  groupBy?: GroupByFields;
}

export interface History {
  task: string;
  title: string;
  url: string;
  filename: string;
  details: string;
  time: string;
  id: number;
}
