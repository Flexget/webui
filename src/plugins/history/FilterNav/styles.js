import styled from '@emotion/styled';
import { css } from '@emotion/core';

import { Icon } from '@material-ui/core';

export const RotatingIcon = styled(Icon)`
  transition: ${({ theme }) => theme.transitions.create()};
  transform: ${({ rotate }) => rotate && 'rotate(180deg)'};
`;

export const Padded = styled.div`
  padding: 0.5rem 2rem 0.5rem;
`;

export const textField = css`
  width: 7.5rem;
`;
