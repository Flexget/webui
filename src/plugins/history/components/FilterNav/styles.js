import styled from 'react-emotion';
import Icon from '@material-ui/core/Icon';
import TextField from '@material-ui/core/TextField';
import theme from 'theme';

export const RotatingIcon = styled(Icon)`
  transition: ${theme.transitions.create()};
  transform: ${({ rotate }) => rotate && 'rotate(180deg)'};
`;

export const PaddedTextField = styled(TextField)`
  margin: 0.5rem 2rem 0.5rem;
`;
