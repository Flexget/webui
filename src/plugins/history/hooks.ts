import { Reducer, useReducer, useEffect } from 'react';
import { stringify } from 'qs';
import { useFlexgetAPI } from 'core/api';
import { HistoryItem, GetHistoryOptions } from './types';

interface State {
  items: HistoryItem[];
  total: number;
}

interface Action {
  items: HistoryItem[];
  headers: Headers;
  reset: boolean;
}

const historyReducer: Reducer<State, Action> = (state, { items, headers, reset }) => {
  const total = parseInt(headers.get('total-count') ?? '0', 10);
  if (reset) {
    return {
      items,
      total,
    };
  }
  return {
    items: [...state.items, ...items],
    total,
  };
};

export const useHistoryPlugin = (options: GetHistoryOptions) => {
  const [state, dispatch] = useReducer(historyReducer, { items: [], total: 0 });

  const [requestState, request] = useFlexgetAPI<HistoryItem[]>(`/history?${stringify(options)}`);

  useEffect(() => {
    const fn = async () => {
      const resp = await request();
      if (resp.ok) {
        dispatch({ items: resp.data, headers: resp.headers, reset: options.page === 1 });
      }
    };
    fn();
  }, [options.page, request]);

  return { ...state, ...requestState };
};
