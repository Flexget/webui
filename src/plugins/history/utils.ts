import { groupBy } from 'utils/array';
import { HistoryItem, GroupByFields } from './types';

export const getGroupedHistory = (items: HistoryItem[], grouping: GroupByFields) =>
  groupBy(
    items,
    grouping,
    grouping === GroupByFields.Time ? (i: string) => i.split('T')[0] : undefined,
  );
