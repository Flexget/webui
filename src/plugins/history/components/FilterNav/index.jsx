import React, { Component } from 'react';
import PropTypes from 'prop-types';
import IconButton from 'material-ui/IconButton';
import { MenuItem } from 'material-ui/Menu';
import SecondaryNav from 'components/SecondaryNav';
import { Spacer } from 'components/styles';
import { RotatingIcon, PaddedTextField } from './styles';

const ENTER_KEY = 13;
const groupByFields = [
  {
    value: 'task',
    label: 'Task',
  },
  {
    value: 'time',
    label: 'Time',
  },
];

const sortByFields = [
  {
    value: 'details',
    label: 'Details',
  },
  {
    value: 'filename',
    label: 'Filename',
  },
  {
    value: 'id',
    label: 'ID',
  },
  {
    value: 'task',
    label: 'Task',
  },
  {
    value: 'time',
    label: 'Time',
  },
  {
    value: 'title',
    label: 'Title',
  },
  {
    value: 'url',
    label: 'URL',
  },
];

class FilterNav extends Component {
  static propTypes = {
    order: PropTypes.oneOf(['asc', 'desc']).isRequired,
    grouping: PropTypes.oneOf(groupByFields.map(({ value }) => value)).isRequired,
    sort: PropTypes.oneOf(sortByFields.map(({ value }) => value)).isRequired,
    toggleOrder: PropTypes.func.isRequired,
    handleChange: PropTypes.func.isRequired,
  };

  state = {
    task: '',
  }

  handleChange = (event) => {
    this.setState({ task: event.target.value });
  }

  handleKeyPress = (event) => {
    const { handleChange } = this.props;
    if (event.which === ENTER_KEY) {
      handleChange('task')(event);
    }
  }

  render() {
    const {
      order,
      grouping,
      sort,
      toggleOrder,
      handleChange,
    } = this.props;

    const {
      task,
    } = this.state;

    return (
      <SecondaryNav>
        <PaddedTextField
          label="Filter By Task"
          id="task-filter"
          placeholder="Task Name"
          value={task}
          onChange={this.handleChange}
          onKeyPress={this.handleKeyPress}
        />
        <Spacer />
        <PaddedTextField
          id="sort-by"
          select
          label="Sort By"
          value={sort}
          onChange={handleChange('sort')}
        >
          {sortByFields.map(({ value, label }) => (
            <MenuItem value={value} key={value}>{label}</MenuItem>
          ))}
        </PaddedTextField>
        <PaddedTextField
          id="group-by"
          label="Group By"
          select
          value={grouping}
          onChange={handleChange('grouping')}
        >
          {groupByFields.map(({ value, label }) => (
            <MenuItem value={value} key={value}>{label}</MenuItem>
          ))}
        </PaddedTextField>
        <IconButton color="inherit" onClick={toggleOrder}>
          <RotatingIcon rotate={order === 'desc'} className="fa fa-chevron-up" />
        </IconButton>
      </SecondaryNav>
    );
  }
}

export default FilterNav;
