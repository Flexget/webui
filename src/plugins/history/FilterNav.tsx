import React, { FC, useEffect } from 'react';
import { Toolbar, Tooltip } from '@material-ui/core';
import { css } from '@emotion/core';
import SelectField from 'common/inputs/formik/SelectField';
import { Spacer } from 'common/styles';
import { useFormikContext } from 'formik';
import { useDebounce } from 'utils/hooks';
import DirectionButton from 'common/inputs/formik/DirectionButton';
import { TaskContainer } from 'core/tasks/hooks';
import { useContainer } from 'unstated-next';
import { GroupByFields, GetHistoryOptions, SortByFields } from './types';

const fieldWrapper = css`
  padding: 0.5rem 2rem 0.5rem;
`;

const textField = css`
  width: 7.5rem;
`;

const selectField = css`
  min-width: 20rem;
`;
const groupByFields = [
  {
    value: GroupByFields.Task,
    label: 'Task',
  },
  {
    value: GroupByFields.Time,
    label: 'Time',
  },
];

const sortByFields = [
  {
    value: SortByFields.Details,
    label: 'Details',
  },
  {
    value: SortByFields.Filename,
    label: 'Filename',
  },
  {
    value: SortByFields.Id,
    label: 'ID',
  },
  {
    value: SortByFields.Task,
    label: 'Task',
  },
  {
    value: SortByFields.Time,
    label: 'Time',
  },
  {
    value: SortByFields.Title,
    label: 'Title',
  },
  {
    value: SortByFields.Url,
    label: 'URL',
  },
];

const FilterNav: FC = () => {
  const { values, submitForm } = useFormikContext<GetHistoryOptions>();

  const debouncedValues = useDebounce(values);
  const { tasks } = useContainer(TaskContainer);

  useEffect(() => {
    submitForm();
  }, [...Object.values(debouncedValues), submitForm]);

  return (
    <Toolbar>
      <div css={fieldWrapper}>
        <SelectField
          css={selectField}
          autoFocus
          id="task"
          label="Task Name"
          name="task"
          options={tasks.map(({ name }) => ({ value: name, label: name }))}
        />
      </div>
      <Spacer />
      <div css={fieldWrapper}>
        <SelectField id="sort" name="sort" label="Sort By" options={sortByFields} css={textField} />
      </div>
      <div>
        <SelectField
          id="grouping"
          label="Group By"
          name="grouping"
          options={groupByFields}
          css={textField}
        />
      </div>
      <Tooltip title="Toggle Direction">
        <DirectionButton color="inherit" name="order" id="order" aria-label="Toggle Direction" />
      </Tooltip>
    </Toolbar>
  );
};

export default FilterNav;
