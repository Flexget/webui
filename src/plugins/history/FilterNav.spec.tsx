import React from 'react';
import { Formik } from 'formik';
import fetchMock from 'fetch-mock';
import renderer from 'react-test-renderer';
import { Direction } from 'utils/query';
import { BaseProviders } from 'utils/tests';
import { TaskContainer } from 'plugins/tasks/hooks';
import { GroupByFields, SortByFields } from './types';
import FilterNav from './FilterNav';

describe('plugins/history/FilterNav', () => {
  beforeEach(() => {
    fetchMock.get('/api/tasks', []).catch();
  });

  afterEach(() => {
    fetchMock.reset();
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(
        <BaseProviders>
          <TaskContainer.Provider>
            <Formik
              initialValues={{
                grouping: GroupByFields.Time,
                task: '',
                page: 1,
                order: Direction.Desc,
                sort: SortByFields.Time,
              }}
              onSubmit={jest.fn()}
            >
              <FilterNav />
            </Formik>
          </TaskContainer.Provider>
        </BaseProviders>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
