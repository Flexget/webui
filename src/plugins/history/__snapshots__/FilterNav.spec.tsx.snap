// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins/history/FilterNav renders correctly 1`] = `
.emotion-0 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

<div
  className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
>
  <div
    css={
      Object {
        "map": undefined,
        "name": "vm19w5",
        "next": undefined,
        "styles": "
  padding: 0.5rem 2rem 0.5rem;
",
        "toString": [Function],
      }
    }
  >
    <div
      className="MuiFormControl-root MuiTextField-root"
      css={
        Object {
          "map": undefined,
          "name": "tr6zgo",
          "next": undefined,
          "styles": "
  min-width: 20rem;
",
          "toString": [Function],
        }
      }
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
        data-shrink={false}
        htmlFor="task"
        id="task-label"
      >
        Task Name
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="task-label task"
          className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
          id="task"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "&#8203;",
              }
            }
          />
        </div>
        <input
          autoFocus={true}
          name="task"
          onAnimationStart={[Function]}
          type="hidden"
          value=""
        />
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root MuiSelect-icon"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
  </div>
  <div
    className="emotion-0"
  />
  <div
    css={
      Object {
        "map": undefined,
        "name": "vm19w5",
        "next": undefined,
        "styles": "
  padding: 0.5rem 2rem 0.5rem;
",
        "toString": [Function],
      }
    }
  >
    <div
      className="MuiFormControl-root MuiTextField-root"
      css={
        Object {
          "map": undefined,
          "name": "1v9q0yb",
          "next": undefined,
          "styles": "
  width: 7.5rem;
",
          "toString": [Function],
        }
      }
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
        data-shrink={true}
        htmlFor="sort"
        id="sort-label"
      >
        Sort By
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="sort-label sort"
          className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
          id="sort"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          Time
        </div>
        <input
          autoFocus={false}
          name="sort"
          onAnimationStart={[Function]}
          type="hidden"
          value="time"
        />
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root MuiSelect-icon"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
  </div>
  <div>
    <div
      className="MuiFormControl-root MuiTextField-root"
      css={
        Object {
          "map": undefined,
          "name": "1v9q0yb",
          "next": undefined,
          "styles": "
  width: 7.5rem;
",
          "toString": [Function],
        }
      }
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
        data-shrink={true}
        htmlFor="grouping"
        id="grouping-label"
      >
        Group By
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="grouping-label grouping"
          className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
          id="grouping"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          Time
        </div>
        <input
          autoFocus={false}
          name="grouping"
          onAnimationStart={[Function]}
          type="hidden"
          value="time"
        />
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root MuiSelect-icon"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
  </div>
  <button
    aria-describedby={null}
    aria-label="Toggle Direction"
    className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
    css={[Function]}
    disabled={false}
    id="order"
    name="order"
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    title="Toggle Direction"
    type="button"
  >
    <span
      className="MuiIconButton-label"
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
        />
      </svg>
    </span>
    <span
      className="MuiTouchRipple-root"
    />
  </button>
</div>
`;
